### ISSUE TITLE
Perform hyperparameter tuning and cross validation on L2G

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#3253]
As a **developer**, I want **to run cross-validation and hyperparameter tuning on L2G** because **this will allow us to enhance the model performance**.

## Background

The L2G model is based on a gradient boosting (XGBoost) algorithm that typically benefits from careful tuning of hyperparameters to obtain an optimal performance.
**Our current set up allows us to parametrise variables** like the depth of the tree, learning rate, or to apply regularisation values just by defining the parameter grid through the configuration file.
At the same time, with W&B we will be able to evaluate how the model is reacting to the different parameter combination.

Personally, I don't expect these will play a huge role in the performance of L2G, but I hope to see subtle improvements in model accuracy and the ability to generalise.

## Tasks

- [ ] Run cross validation on a grid of hyperparameters. The model will be trained and tested for each possible combination of hyperparameters
- [ ] Analyse the results in W&B to understand how specific variables impact the performance of the model
- [ ] Update the L2G step configuration with the set of parameters yields the best results

## Notes
The work describe in https://github.com/opentargets/issues/issues/3252 will have an impact on this issue. I'd suggest attacking that one first.

**Comment by addramir** [comment]
Is it done?

**Comment by ireneisdoomed** [comment]
@addramir No but I think we should. That's why I asked your notebook of OOV, to incorporate the approach.

**Comment by ireneisdoomed** [comment]
I have implemented a cross validation strategy to have a more credible model evaluation (fixed set of hyperparameters). This is a common practice in ML, and in our case I think it is particularly necessary because we constantly see that training metrics in W&B
look really good even though the actual scores for the full prediction set is not good.

For this I have split the large v5.1 gold standard (`gs://genetics-portal-dev-analysis/yt4/20241024_EGL_playground/training_set/v5_1_full.json`) into train, eval, and test. This dataset has 10598 positives and 145702 negatives.

My intention was to measure:
- More accurate evaluation metrics that is not based on a single fold evaluation run
- Overfitting
- How large should our training set be

## Naive cross validation is useless
```
Training fold 1/5
Fold 1 AUC: 0.973
Training fold 2/5
Fold 2 AUC: 0.973
Training fold 3/5
Fold 3 AUC: 0.972
Training fold 4/5
Fold 4 AUC: 0.974
Training fold 5/5
Fold 5 AUC: 0.973

Cross-validation results:
AUC: 0.973 ± 0.001
```
Because our data is indexed by EFO/Gene pairs, and then exploded to credible sets, we should never test on any EFO/gene that is part of the training set. The data leakage is obvious, but it was nice to measure it.
**We should split train and test to take this into account.**

## Cross validation results stratifying by EFO/Gene pairs
Results after creating the folds so that each row for the same EFO/gene stay together in either training or test set look less homogenous and more realistic (AUC 0.971 ± 0.006).

However, metrics were still too good. Learning curves are good to diagnose overfitting. They show model performance on the train and validation datasets over time. These are learning curves for precision and for average precision (AUPR):

![Image](https://github.com/user-attachments/assets/4ba8ff3c-11ab-4c6c-bf6f-a1d45bc4b6cc)

- Training scores are consistently higher than validation scores, indicating **overfitting**
- Both curves are very similar, but the overfitting is slightly higher for the average precision, which is a more robuts metric.
- Learning doesn't increase at around 60% of the data
- We have very high training precisions at the beginning, and as the training size grows, the precision decreases. This, together with the inverse behaviour for the evaluation precision, indicates that the model is able to memorise the data when the training set is very small, and as the training set grows, patterns are less obvious.

---
I don't know if we can significantly improve this, but we can test whether playing with the model parameters a bit and doing regularization improves the overfitting.
cc @addramir

**Comment by addramir** [comment]
Looks interesting, thank you! It is definitely worth trying to do regularisation if it is possible. Especially because the training set we are going to use is only ~450 GSPs that is much less than ~3000 needed for proper training (30% of the data). In any cae, if it doesn't work - we will proceed as is.

**Comment by ireneisdoomed** [comment]
## Decreasing overfitting
**Regularization** has improved decreasing the gap between the training and eval metrics and in general getting a better average precision in the evaluation set.

The principle is to essentially introduce some penalty so that the model doesn't focus too much on noise in the data. The typical techniques is to tweak the alpha value (L1 regularization) or lambda (L2). However, the estimator we are using doesn't have those 2 parameters specifically.

I have implemented an approach tweaking `learning_rate` and `subsample` following [this](https://labex.io/tutorials/ml-gradient-boosting-regularization-49154) and [this](https://www.datacamp.com/tutorial/guide-to-the-gradient-boosting-algorithm) tutorials.

**Learning rate**: is a multiplier that shrinks the contribution of each tree to the overall model. Theoretically lower values (e.g., 0.1 or 0.2) can help prevent overfitting. It's 0.1 by default.
**Subsampling:** representes the fraction of samples to be used to fit each tree. Lower values (e.g., 0.5) can also help prevent overfitting. It's 1 by default.
**Max depth**: controls the number of levels in each decision tree. The deeper the tree, the more complex and computationally expensive the model becomes.
**CCP Alpha:** Cost complexity pruning provides another option to control the size of a tree. Greater values of ccp_alpha increase the number of nodes pruned.
**Minimum samples split:** controls how branches split in decision trees. A low value for the number of samples in leaf nodes makes the overall algorithm sensitive to noise.
**Minimum samples leaf:** The minimum number of samples required to split an internal node. A larger minimum number of samples helps to prevent overfitting by making it more difficult for the trees to create splits based on too few data points.


### Cross validation results
Based on the above, I've run 3-fold cross validation for the combination of each of these parameters.
```
param_grid = ParameterGrid({
    'learning_rate': [0.1],
    'subsample': [0.5, 1.0],
    'ccp_alpha': [0.0],
    'min_samples_leaf': [5, 10],
    'min_samples_split': [2, 5, 10],
    "max_depth": [10]
})
```

👉 **Results of all 432 combinations are here, ordered by average precision in the CV folds and average precision in the holdout set** https://docs.google.com/spreadsheets/d/1R4RD1PFCwRxXvZ_2JBTXbfAQD5EnM8xgqqK7sRa38CA/edit?usp=sharing

### Best model validation

These learning curves are analogous to the one showed in the previous comment, only that here I am plotting again the average precision estimates for the top 12 models.

![Image](https://github.com/user-attachments/assets/521ea5dd-c0a1-43dd-ad65-036cd06b50bc)

Both the plot and the table show that metrics are practically the same, so I've decided to choose Run ID 27 (`{'ccp_alpha': 0.0, 'learning_rate': 0.1, 'max_depth': 10, 'min_samples_leaf': 5, 'min_samples_split': 5, 'subsample': 1.0}`) as the best model.

In the learning curve we also see that with more than 60% of the training set the gain in performance is minimal. However, based on what I have been reading, _this doesn't mean that we should reduce the amount of training data as long as overfitting doesn't increase._ 

This is the performance during cross validation and on the held out test.
```
Performing cross-validation...
Fold 1: Average Precision = 0.879
Fold 2: Average Precision = 0.814
Fold 3: Average Precision = 0.856
Fold 4: Average Precision = 0.732
Fold 5: Average Precision = 0.730

Cross-validation Average Precision: 0.802 (+/- 0.124)

Training final model on full training set...

Evaluating on held-out test set...
Test Set Average Precision: 0.759
```

**Numbers don't look as promising as the ones above, but the methodology followed here is much better.** In fact, I see a much better performance in terms of selectivity as well:
```
1 per cs: 345498  ---> 45k more than current best model
 2 per cs: 18619  ---> 11k less than current best model
 >2 per cs: 2227   ---> 2k less than current best model
```
(more validation metrics in the first sheet of the [spreadsheet](https://docs.google.com/spreadsheets/d/1R4RD1PFCwRxXvZ_2JBTXbfAQD5EnM8xgqqK7sRa38CA/edit?usp=sharing) (my best model is `fm0/gs_v1_full_best_cv`)

## New training approach including out of sample validation

I trust these results more based on these 2 key improvements:
- **Train/Test split is based on EFO/Gene pairs to avoid data leakage**.  This is done both to generate the splits during cross validation and to generate the hold out test. This means that the test set will have completely new gene/trait associations, so we can consider it a valid out of sample testing set.
- **Cross validation just gives you a less biased picture**.

In general, we have implemented an algorithm that includes a much larger gold standard (10598 positives and 145702 negatives) that we split in a way where we account for overfitting, and where the out of sample validation is part of the pipeline. The diagram illustrates the idea.
![Image](https://github.com/user-attachments/assets/5cf9d870-c377-422e-bfb6-9653b76caa45)
